File: src/xdevs/core/modeling/Atomic.java
/*
 * Copyright (C) 2014-2015 José Luis Risco Martín <jlrisco@ucm.es> and 
 * Saurabh Mittal <smittal@duniptech.com>.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, see
 * http://www.gnu.org/licenses/
 *
 * Contributors:
 *  - José Luis Risco Martín
 */
package xdevs.core.modeling;

import xdevs.core.util.Constants;

/**
 *
 * @author José L. Risco Martín and Saurabh Mittal
 */
public abstract class Atomic extends Component {

    // DevsAtomic attributes
    protected String phase = Constants.PHASE_PASSIVE;
    protected double sigma = Constants.INFINITY;

    public Atomic(String name) {
        super(name);
    }

    public Atomic() {
        this(Atomic.class.getSimpleName());
    }

    // DevsAtomic methods
    public double ta() {
        return sigma;
    }

    public abstract void deltint();

    public abstract void deltext(double e);

    public void deltcon(double e) {
        deltint();
        deltext(0);
    }

    public abstract void lambda();

    public void resume(double e) {
        sigma = sigma - e;
    }

    public void holdIn(String phase, double sigma) {
        this.phase = phase;
        this.sigma = sigma;
    }

    public void activate() {
        this.phase = Constants.PHASE_ACTIVE;
        this.sigma = 0;
    }

    public void passivate() {
        this.phase = Constants.PHASE_PASSIVE;
        this.sigma = Constants.INFINITY;
    }

    public void passivateIn(String phase) {
        this.phase = phase;
        this.sigma = Constants.INFINITY;
    }

    public boolean phaseIs(String phase) {
        return this.phase.equals(phase);
    }

    public String getPhase() {
        return phase;
    }

    public final void setPhase(String phase) {
        this.phase = phase;
    }

    public double getSigma() {
        return sigma;
    }

    public final void setSigma(double sigma) {
        this.sigma = sigma;
    }

    public String showState() {
        StringBuilder sb = new StringBuilder(name + ":[");
        sb.append("\tstate: ").append(phase);
        sb.append("\t, sigma: ").append(sigma);
        sb.append("]");
        return sb.toString();
    }

    public String toXml() {
        StringBuilder builder = new StringBuilder();
        StringBuilder tabs = new StringBuilder();
        Component parent = this.parent;
        int level = 0;
        while (parent!=null) {
            tabs.append("\t");
            parent = parent.parent;
            level++;
        }
        builder.append(tabs).append("<atomic name=\"").append(this.getName()).append("\"");
        builder.append(" class=\"").append(this.getClass().getCanonicalName()).append("\"");
        builder.append(" host=\"127.0.0.1\"");
        builder.append(" port=\"").append(5000 + level).append("\"");
        builder.append(">\n");
        builder.append(tabs).append("</atomic>\n");

        return builder.toString();
    }
}

File: modeling/atomic.go
package modeling

import (
	"fmt"
	"github.com/pointlesssoft/godevs/util"
)

type Atomic interface {
	Component
	TA() float64
	DeltInt()
	DeltExt(e float64)
	DeltCon(e float64)
	Lambda()
	HoldIn(phase string, sigma float64)
	Activate()
	Passivate()
	PassivateIn(phase string)
	PhaseIs(phase string) bool
	GetPhase() string
	SetPhase(phase string)
	GetSigma() float64
	SetSigma(sigma float64)
	ShowState() string
}

func NewAtomic(name string) Atomic {
	a := atomic{NewComponent(name), util.PASSIVE, util.INFINITY}
	return &a
}

type atomic struct {
	Component
	phase string
	sigma float64
}

func (a *atomic) TA() float64 {
	return a.sigma
}

func (a *atomic) DeltInt() {
	panic("implement me")
}

func (a *atomic) DeltExt(e float64) {
	panic("implement me")
}

func (a *atomic) DeltCon(e float64) {
	a.DeltInt()
	a.DeltExt(0)
}

func (a *atomic) Lambda() {
	panic("implement me")
}

func (a *atomic) HoldIn(phase string, sigma float64) {
	a.phase = phase
	a.sigma = sigma
}

func (a *atomic) Activate() {
	a.phase = util.ACTIVE
	a.sigma = 0
}

func (a *atomic) Passivate() {
	a.phase = util.PASSIVE
	a.sigma = util.INFINITY
}

func (a *atomic) PassivateIn(phase string) {
	a.phase = phase
	a.sigma = util.INFINITY
}

func (a *atomic) PhaseIs(phase string) bool {
	return a.phase == phase
}

func (a *atomic) GetPhase() string {
	return a.phase
}

func (a *atomic) SetPhase(phase string) {
	a.phase = phase
}

func (a *atomic) GetSigma() float64 {
	return a.sigma
}

func (a *atomic) SetSigma(sigma float64) {
	a.sigma = sigma
}

func (a *atomic) ShowState() string {
	state := a.GetName() + " [ "
	state += "\tstate: " + a.phase
	state += "\tsigma: " + fmt.Sprintf("%f", a.sigma)
	state += " ]"
	return state
}
